File: Medium/1367-Linked List in Binary Tree/Recursion.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {ListNode} head
# @param {TreeNode} root
# @return {Boolean}
def is_sub_path(head, root)
  return false if root.nil?
  return true if head.val == root.val && check_path(head, root)

  is_sub_path(head, root.left) || is_sub_path(head, root.right)
end

# @param {ListNode} head
# @param {TreeNode} root
# @return {Boolean}
def check_path(head, root)
  return true if head.nil?
  return false if root.nil? || head.val != root.val

  check_path(head.next, root.left) || check_path(head.next, root.right)
end

File: Problemset/1367-Linked List in Binary Tree/Recursion.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubPath(self, head: ListNode, root: TreeNode) -> bool:
        if root is None:
            return False
        elif head.val == root.val and self.checkPath(head, root):
            return True

        return self.isSubPath(head, root.left) or self.isSubPath(head, root.right)

    def checkPath(self, head: ListNode, root: TreeNode) -> bool:
        if head is None:
            return True
        elif root is None or head.val != root.val:
            return False

        return self.checkPath(head.next, root.left) or self.checkPath(head.next, root.right)
